### ISSUE TITLE
`with_mut` is unsound.

### ISSUE THREAD

**Issue by joshua-maros** [joshua-maros/ouroboros#98]
Example:

```rust
use ouroboros::self_referencing;

pub struct PrintStrRef<'a>(&'a str);

impl Drop for PrintStrRef<'_> {
    fn drop(&mut self) {
        println!("Dropping {}", self.0);
    }
}

#[self_referencing]
pub struct Tricky {
    data1: String,
    #[borrows(data1)]
    #[covariant]
    ref1: PrintStrRef<'this>,
    data2: String,
    #[borrows(data2)]
    #[covariant]
    ref2: PrintStrRef<'this>,
}

fn main() {
    let mut t = Tricky::new(
        "A".to_owned(),
        |a| PrintStrRef(a),
        "B".to_owned(),
        |b| PrintStrRef(b),
    );
    t.with_mut(|fields| {
        *fields.ref1 = PrintStrRef(fields.data2);
    });
    drop(t);
}
```
`data2` is dropped before `ref1`, leading to undefined behavior.

**Comment by joshua-maros** [comment]
I think this can be fixed by adding more fake lifetimes to the `BorrowedMutFields` struct.

**Comment by joshua-maros** [comment]
This is now fixed in 0.18.0.
